<a href="https://colab.research.google.com/github/Julianamihailova/laba/blob/main/lab1/%D0%97%D0%B0%D0%B4%D0%B0%D0%BD%D0%B8%D0%B5_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Задание по практике 1

##Загрузка датасета
В файле classification datasets расположены ссылки на датасеты. Выберите один в соответствии со своим вариантом, скачайте и прочитайте файл

**Не забудьте загрузить нужные библиотеки**

Например: pip3 install pandas

In [7]:
import pandas as pd

In [9]:
df = pd.read_csv("/content/cwurData.csv")
df[:5]

world_rank                            institution         country  \
0           1                     Harvard University             USA   
1           2  Massachusetts Institute of Technology             USA   
2           3                    Stanford University             USA   
3           4                University of Cambridge  United Kingdom   
4           5     California Institute of Technology             USA   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              1                     7                  9                   1   
1              2                     9                 17                   3   
2              3                    17                 11                   5   
3              1                    10                 24                   4   
4              4                     2                 29                   7   

   publications  influence  citations  broad_impact  patents   score  year  
0             1          1          1           NaN        5  100.00  2012  
1            12          4          4           NaN        1   91.67  2012  
2             4          2          2           NaN       15   89.50  2012  
3            16         16         11           NaN       50   86.17  2012  
4            37         22         22           NaN       18   85.21  2012

##Первое знакомство с датасетом
В следующих ячейках выведите:


1.   Колонки датасета
2.   Первые 5 значений
3.   Последние 5 значений
4.   Статистику по числовым колонкам



In [11]:
df.columns



Index(['world_rank', 'institution', 'country', 'national_rank',
       'quality_of_education', 'alumni_employment', 'quality_of_faculty',
       'publications', 'influence', 'citations', 'broad_impact', 'patents',
       'score', 'year'],
      dtype='object')

In [12]:
df[:5]

world_rank                            institution         country  \
0           1                     Harvard University             USA   
1           2  Massachusetts Institute of Technology             USA   
2           3                    Stanford University             USA   
3           4                University of Cambridge  United Kingdom   
4           5     California Institute of Technology             USA   

   national_rank  quality_of_education  alumni_employment  quality_of_faculty  \
0              1                     7                  9                   1   
1              2                     9                 17                   3   
2              3                    17                 11                   5   
3              1                    10                 24                   4   
4              4                     2                 29                   7   

   publications  influence  citations  broad_impact  patents   score  year  
0             1          1          1           NaN        5  100.00  2012  
1            12          4          4           NaN        1   91.67  2012  
2             4          2          2           NaN       15   89.50  2012  
3            16         16         11           NaN       50   86.17  2012  
4            37         22         22           NaN       18   85.21  2012

In [13]:
df.tail()

world_rank                      institution   country  national_rank  \
2195         996        University of the Algarve  Portugal              7   
2196         997            Alexandria University     Egypt              4   
2197         998      Federal University of Ceará    Brazil             18   
2198         999           University of A Coruña     Spain             40   
2199        1000  China Pharmaceutical University     China             83   

      quality_of_education  alumni_employment  quality_of_faculty  \
2195                   367                567                 218   
2196                   236                566                 218   
2197                   367                549                 218   
2198                   367                567                 218   
2199                   367                567                 218   

      publications  influence  citations  broad_impact  patents  score  year  
2195           926        845        812         969.0      816  44.03  2015  
2196           997        908        645         981.0      871  44.03  2015  
2197           830        823        812         975.0      824  44.03  2015  
2198           886        974        812         975.0      651  44.02  2015  
2199           861        991        812         981.0      547  44.02  2015

In [14]:
df.describe()

world_rank  national_rank  quality_of_education  alumni_employment  \
count  2200.000000    2200.000000           2200.000000        2200.000000   
mean    459.590909      40.278182            275.100455         357.116818   
std     304.320363      51.740870            121.935100         186.779252   
min       1.000000       1.000000              1.000000           1.000000   
25%     175.750000       6.000000            175.750000         175.750000   
50%     450.500000      21.000000            355.000000         450.500000   
75%     725.250000      49.000000            367.000000         478.000000   
max    1000.000000     229.000000            367.000000         567.000000   

       quality_of_faculty  publications    influence    citations  \
count         2200.000000   2200.000000  2200.000000  2200.000000   
mean           178.888182    459.908636   459.797727   413.417273   
std             64.050885    303.760352   303.331822   264.366549   
min              1.000000      1.000000     1.000000     1.000000   
25%            175.750000    175.750000   175.750000   161.000000   
50%            210.000000    450.500000   450.500000   406.000000   
75%            218.000000    725.000000   725.250000   645.000000   
max            218.000000   1000.000000   991.000000   812.000000   

       broad_impact      patents        score         year  
count   2000.000000  2200.000000  2200.000000  2200.000000  
mean     496.699500   433.346364    47.798395  2014.318182  
std      286.919755   273.996525     7.760806     0.762130  
min        1.000000     1.000000    43.360000  2012.000000  
25%      250.500000   170.750000    44.460000  2014.000000  
50%      496.000000   426.000000    45.100000  2014.000000  
75%      741.000000   714.250000    47.545000  2015.000000  
max     1000.000000   871.000000   100.000000  2015.000000

**Ответьте на следующие вопросы:**


1.   Какие средние значения в каждой числовой колонке вашего датасета?
2.   Минимальные и максимальные значения в каждой числовой колонке вашего датасета





1. 459.590909	40.278182	275.100455	357.116818	178.888182	459.908636	459.797727	413.417273	496.699500	433.346364	47.798395	2014.318182
2.   min - 1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	1.000000	43.360000	2012.000000
max - 1000.000000	229.000000	367.000000	567.000000	218.000000	1000.000000	991.000000	812.000000	1000.000000	871.000000	100.000000	2015.000000



##Типизация
Присвойте соответствующим колонкам один из типов (category, float, int)

In [17]:
"""

df.Series = df.Series.astype("category")
df.X = df.X.astype("Float64")
df.Y = df.Y.astype("Float64")

"""
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2200 entries, 0 to 2199
Data columns (total 14 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   world_rank            2200 non-null   int64  
 1   institution           2200 non-null   object 
 2   country               2200 non-null   object 
 3   national_rank         2200 non-null   int64  
 4   quality_of_education  2200 non-null   int64  
 5   alumni_employment     2200 non-null   int64  
 6   quality_of_faculty    2200 non-null   int64  
 7   publications          2200 non-null   int64  
 8   influence             2200 non-null   int64  
 9   citations             2200 non-null   int64  
 10  broad_impact          2000 non-null   float64
 11  patents               2200 non-null   int64  
 12  score                 2200 non-null   float64
 13  year                  2200 non-null   int64  
dtypes: float64(2), int64(10), object(2)
memory usage: 240.8+ KB


Какие признаки вы можете выделить в своем датасете (категориальные, числовые и др., **минимум 3 разных**, помните что можно также выделить признаки более точно, например числовые - порядковые, количественные и т.д.)?

##Фильтрация
Офильтруйте 3 колонки вашего датасета и выведите результат

In [18]:
df[df["quality_of_education"]>200]

world_rank                                     institution      country  \
223           24                       Seoul National University  South Korea   
226           27         University of California, San Francisco          USA   
233           34                                 Keio University        Japan   
239           40                               Waseda University        Japan   
245           46  Pennsylvania State University, University Park          USA   
...          ...                                             ...          ...   
2195         996                       University of the Algarve     Portugal   
2196         997                           Alexandria University        Egypt   
2197         998                     Federal University of Ceará       Brazil   
2198         999                          University of A Coruña        Spain   
2199        1000                 China Pharmaceutical University        China   

      national_rank  quality_of_education  alumni_employment  \
223               1                   355                  9   
226              20                   355                478   
233               3                   271                  5   
239               4                   355                  7   
245              29                   226                 42   
...             ...                   ...                ...   
2195              7                   367                567   
2196              4                   236                566   
2197             18                   367                549   
2198             40                   367                567   
2199             83                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
223                  210            38        165         87         107.0   
226                   23            24          8         17           9.0   
233                  210           299        243        310         266.0   
239                  210           391        384        283         380.0   
245                   65            33         36         26          33.0   
...                  ...           ...        ...        ...           ...   
2195                 218           926        845        812         969.0   
2196                 218           997        908        645         981.0   
2197                 218           830        823        812         975.0   
2198                 218           886        974        812         975.0   
2199                 218           861        991        812         981.0   

      patents  score  year  
223         5  66.06  2014  
226        57  63.36  2014  
233       159  59.84  2014  
239       204  58.69  2014  
245       100  57.19  2014  
...       ...    ...   ...  
2195      816  44.03  2015  
2196      871  44.03  2015  
2197      824  44.03  2015  
2198      651  44.02  2015  
2199      547  44.02  2015  

[1600 rows x 14 columns]

In [20]:
df[df["patents"]<55]

world_rank                                  institution         country  \
0              1                           Harvard University             USA   
1              2        Massachusetts Institute of Technology             USA   
2              3                          Stanford University             USA   
3              4                      University of Cambridge  United Kingdom   
4              5           California Institute of Technology             USA   
...          ...                                          ...             ...   
1364         165  Pohang University of Science and Technology     South Korea   
1373         174              Technical University of Denmark         Denmark   
1377         178                          Hokkaido University           Japan   
1391         192                           Hanyang University     South Korea   
1509         310  Gwangju Institute of Science and Technology     South Korea   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
3                 1                    10                 24   
4                 4                     2                 29   
...             ...                   ...                ...   
1364              5                   367                567   
1373              3                   351                510   
1377             10                   279                277   
1391              6                   367                168   
1509              9                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1           NaN   
1                      3            12          4          4           NaN   
2                      5             4          2          2           NaN   
3                      4            16         16         11           NaN   
4                      7            37         22         22           NaN   
...                  ...           ...        ...        ...           ...   
1364                 218           285        355        212         281.0   
1373                 202           220        230        161         216.0   
1377                 135           148        173        287         251.0   
1391                 218           286        471        368         396.0   
1509                 218           540        538        368         529.0   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
3          50   86.17  2012  
4          18   85.21  2012  
...       ...     ...   ...  
1364       24   48.24  2015  
1373       46   48.06  2015  
1377       53   47.97  2015  
1391       31   47.60  2015  
1509       54   45.88  2015  

[197 rows x 14 columns]

In [26]:
df[df["country"]=="USA"]

world_rank                            institution country  \
0              1                     Harvard University     USA   
1              2  Massachusetts Institute of Technology     USA   
2              3                    Stanford University     USA   
4              5     California Institute of Technology     USA   
5              6                   Princeton University     USA   
...          ...                                    ...     ...   
2100         901     University of Southern Mississippi     USA   
2105         906                     Oakland University     USA   
2112         913             University of North Dakota     USA   
2128         929         University of Texas at El Paso     USA   
2135         936       University of California, Merced     USA   

      national_rank  quality_of_education  alumni_employment  \
0                 1                     7                  9   
1                 2                     9                 17   
2                 3                    17                 11   
4                 4                     2                 29   
5                 5                     8                 14   
...             ...                   ...                ...   
2100            225                   367                567   
2105            226                   367                567   
2112            227                   367                567   
2128            228                   367                442   
2135            229                   367                567   

      quality_of_faculty  publications  influence  citations  broad_impact  \
0                      1             1          1          1           NaN   
1                      3            12          4          4           NaN   
2                      5             4          2          2           NaN   
4                      7            37         22         22           NaN   
5                      2            53         33         26           NaN   
...                  ...           ...        ...        ...           ...   
2100                 218           913        853        812         850.0   
2105                 218           888        810        812         850.0   
2112                 218           917        731        812         867.0   
2128                 218           910        838        812         906.0   
2135                 218           899        540        812         906.0   

      patents   score  year  
0           5  100.00  2012  
1           1   91.67  2012  
2          15   89.50  2012  
4          18   85.21  2012  
5         101   82.50  2012  
...       ...     ...   ...  
2100      676   44.13  2015  
2105      871   44.13  2015  
2112      606   44.12  2015  
2128      706   44.10  2015  
2135      491   44.09  2015  

[573 rows x 14 columns]

##Группировка
Сгруппируйте датасет по какому-либо категориальному признаку и выведите среднее значение численных колонок (см. метод mean в pandas)

In [29]:
df.groupby('publications').mean(numeric_only="True")

world_rank  national_rank  quality_of_education  \
publications                                                    
1                   1.00           1.00                  2.50   
2                  31.50           6.75                 75.00   
3                  22.75          16.75                 37.50   
4                  13.75          10.75                 20.50   
5                  14.75          11.00                 39.75   
...                  ...            ...                   ...   
996               750.00          34.00                367.00   
997               983.50          19.00                295.50   
998               850.50          17.50                361.00   
999               973.50          37.50                361.00   
1000              825.00          34.50                291.00   

              alumni_employment  quality_of_faculty  influence  citations  \
publications                                                                
1                          3.00                1.00       1.00       1.00   
2                         59.25               50.50      16.25      10.25   
3                         30.50               98.75      17.25       6.25   
4                         66.50               15.00       7.00       5.00   
5                         35.00               15.00       7.00       4.25   
...                         ...                 ...        ...        ...   
996                      567.00              218.00     368.00     645.00   
997                      522.00              214.00     923.00     627.00   
998                      522.50              214.00     578.00     627.00   
999                      522.50              214.00     613.00     627.00   
1000                     522.50              214.00     821.50     722.50   

              broad_impact  patents     score    year  
publications                                           
1                      1.0     4.25  100.0000  2013.5  
2                      8.0   100.75   57.6850  2013.5  
3                     12.0    12.50   64.1125  2013.5  
4                      3.0     5.50   74.4775  2013.5  
5                      4.0    32.25   79.4600  2013.5  
...                    ...      ...       ...     ...  
996                  659.0   871.00   44.3000  2015.0  
997                  964.5   711.50   44.1650  2014.5  
998                  774.5   804.00   44.3250  2014.5  
999                  951.0   649.50   44.1750  2014.5  
1000                 777.0   711.50   44.3450  2014.5  

[987 rows x 11 columns]

##Задание для гениев *
Агрегируйте числовой признак по категориальному признаку с использованием метода **без использования** библиотеки pandas, и вычислите медиану и среднее значение для каждой категории.

In [ ]:
#Агрегация